File: 백준/Gold/13913. 숨바꼭질 4/숨바꼭질 4.cc
#include <bits/stdc++.h>
using namespace std;

#define prev aaa

int n, k, prev[100004], visited[100004];
queue<int> q;
vector<int> v;

int main() {
  cin >> n >> k;

  visited[n] = 1;
  q.push(n);

  while (q.size()) {
    int here = q.front();
    q.pop();

    if (here == k) {
      break;
    }

    for (int next : {here + 1, here - 1, here * 2}) {
      if (next < 0 || next > 100000) continue;
      if (visited[next] == 0 || visited[next] == visited[here] + 1) {
        q.push(next);
        visited[next] = visited[here] + 1;
        prev[next] = here;
      }
    }
  }
  for (int i = k; i != n; i = prev[i]) v.push_back(i);

  v.push_back(n);

  cout << visited[k] - 1 << '\n';
  reverse(v.begin(), v.end());
  for (int i : v) cout << i << ' ';
  return 0;
}

// 최단거리: BFS
// 네방향이 아닌 세방향 BFS
// 이전 위치 추적할 떄(trace) prev, next 쓰면 된다.

File: 백준/Gold/13913. 숨바꼭질 4/숨바꼭질 4.js
const [N, K] = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : './input.txt').toString().trim().split('\n')[0].split(' ').map(Number);

function answer () {
   const queue = [];
   const visited = Array(100004).fill(0);
   const prev = Array(100004).fill(-1);

   queue.push(N);
   visited[N] = 1; 
    let p = 0;

    while(p < queue.length){
        const here = queue[p++];

        for (const there of [here + 1, here - 1, here * 2]){
            if(there < 0 || there > 100000) continue;
            if(visited[there] === 0 || visited[there] === visited[here] + 1){
                queue.push(there);
                visited[there] = visited[here] + 1;
                prev[there] = here;
            }
        }
    }

    console.log(visited[K] - 1);
    
    let a = K;
    let v = [];
    
    while(a !== -1){
        v.push(a);
        a = prev[a];
    }

    console.log(v.reverse().join(' '));
    
}

answer();